sql/ttl: add retry logic for TTL deletions

spilchen opened the pull request:
Previously, when a TTL deletion encountered contention, the KV layer would automatically retry up to 100 times. If all 100 retries failed, the error would propagate up the stack, causing the TTL deletion job to restart.

A problem with this approach is that it will retry the TTL deletion using the same batch size, making it likely that it will hit a retry again.

This change introduces retry logic within the TTL deletion job itself. If the KV auto-retry limit is reached, the delete batch size is halved and retried. This process continues until the batch size reaches a single row, at which point it will fail if the operation still cannot succeed.

By dynamically adjusting the batch size instead of failing outright, this change improves the robustness of TTL-based row deletions, reducing the likelihood of job failures that would otherwise require a full restart.

To support this, a new `retry.Batch` struct was added to the `retry` package. In a follow-up PR, this struct will be used to replace similar retry logic in other areas, such as the backfill and merge phases.

Epic: None
Release note (bug fix): The TTL deletion job now includes a retry mechanism that progressively reduces the batch size when encountering contention. This improves the chances of successful deletion without requiring manual adjustments to TTL knobs.
Closes #125103

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/141953)
<!-- Reviewable:end -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141953)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/141953#-OK8N4PKBoPRkmN97Yx0:-OK8N4PL00lBFH4iOKia:bd6si2r) (no related file):*
nice! this lgtm, but my main question is about adding a TTL-specific test.
___
*[`pkg/util/retry/batch.go` line 19 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141953#-OK8MJFU-oMbHP85MPUT:-OK8MJFU-oMbHP85MPUU:b-kw2xis) ([raw file](https://github.com/cockroachdb/cockroach/blob/82b417793aade4fc1396524b184ed25ceaa32326/pkg/util/retry/batch.go#L19)):*
> ```Go
> // to define the batch operation and can optionally define retry behavior
> // using `IsRetriableError` and `OnRetry`.
> type Batch struct {
> ```

this is a nice utility! i think the audience in #db-engineering would be interested to know about it. up to you if you'd like to solicit broader feedback on it before merging, or if you'd like to just announce this later on.
___
*[`pkg/util/retry/batch.go` line 26 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141953#-OK8KZXF2-Wxb_T2FaiQ:-OK8KZXF2-Wxb_T2FaiR:bq7vzsu) ([raw file](https://github.com/cockroachdb/cockroach/blob/82b417793aade4fc1396524b184ed25ceaa32326/pkg/util/retry/batch.go#L26)):*
> ```Go
> 	totalToProcess int64
> 	// processed is the count of items successfully processed so far.
> 	processed int64
> ```

super nit: could these fields just be `int`s?
___
*[`pkg/util/retry/batch.go` line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141953#-OK8K3ijB4K1HWcciJoc:-OK8K3ijB4K1HWcciJod:bnvr6ib) ([raw file](https://github.com/cockroachdb/cockroach/blob/82b417793aade4fc1396524b184ed25ceaa32326/pkg/util/retry/batch.go#L73)):*
> ```Go
> 
> 		// Handle retriable errors by reducing the batch size and retrying.
> 		if b.IsRetriableError(err) {
> ```

i believe this line means that IsRetriable is not optional, but we might want to clarify the comment above that says users "can optionally define retry behavior using `IsRetriableError` and `OnRetry`"
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 413 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141953#-OK8MxhsDMrX8l3BerAR:-OK8MxhsDMrX8l3BerAS:b9tehsh) ([raw file](https://github.com/cockroachdb/cockroach/blob/82b417793aade4fc1396524b184ed25ceaa32326/pkg/sql/ttl/ttljob/ttljob_processor.go#L413)):*
> ```Go
> 			// retry.Batch to allow retrying with a smaller batch size in case of
> 			// an error.
> 			rb := retry.Batch{
> ```

would it be possible to add a TTL specific test that injects one of these retriable errors?
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 456 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141953#-OK8Fv-gCKQR9w33pVVx:-OK8Fv-gCKQR9w33pVVy:bsbal9p) ([raw file](https://github.com/cockroachdb/cockroach/blob/82b417793aade4fc1396524b184ed25ceaa32326/pkg/sql/ttl/ttljob/ttljob_processor.go#L456)):*
> ```Go
> 				OnRetry: func(err error, nextBatchSize int64) error {
> 					log.Infof(ctx,
> 						"kv auto-retry limit reached, reducing batch size to %d rows. Error: %v",
> ```

super nit on the message: could it say "row level TTL reached auto-retry limit, reducing batch size to ..."?

just so that we know the message is coming from row level TTL
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 463 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141953#-OK8MlYj3AoyIz7xXszD:-OK8MlYj3AoyIz7xXszE:b-lhtgw4) ([raw file](https://github.com/cockroachdb/cockroach/blob/82b417793aade4fc1396524b184ed25ceaa32326/pkg/sql/ttl/ttljob/ttljob_processor.go#L463)):*
> ```Go
> 			// Adjust the batch size if we are on the final batch.
> 			deleteBatchSize = min(deleteBatchSize, numExpiredRows-startRowIdx)
> 			if err := rb.Start(ctx, deleteBatchSize); err != nil {
> ```

super nit: `Execute` may be more clear than `Start` for the function name.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141953)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 413 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141953#-OK8MxhsDMrX8l3BerAR:-OKCx5QREMdg7Y664id3:b-md7ner) ([raw file](https://github.com/cockroachdb/cockroach/blob/82b417793aade4fc1396524b184ed25ceaa32326/pkg/sql/ttl/ttljob/ttljob_processor.go#L413)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

would it be possible to add a TTL specific test that injects one of these retriable errors?
</blockquote></details>

Yes, this is a good idea. I had to make a few additional changes to get the test working. Specifically, I converted SelectBuilder and DeleteBuilder to interfaces instead of concrete structs, which allowed me to create a mock to control the errors during batch processing. Additionally, to track the number of batch retries in the test, I added a new metric (jobs.row_level_ttl.num_delete_batch_retries).
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 456 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141953#-OK8Fv-gCKQR9w33pVVx:-OKBeXqSEDySOQq_HJw7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/82b417793aade4fc1396524b184ed25ceaa32326/pkg/sql/ttl/ttljob/ttljob_processor.go#L456)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit on the message: could it say "row level TTL reached auto-retry limit, reducing batch size to ..."?

just so that we know the message is coming from row level TTL
</blockquote></details>

Done.
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 463 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141953#-OK8MlYj3AoyIz7xXszD:-OKBemue8bnSTe9m3gzX:b-egv461) ([raw file](https://github.com/cockroachdb/cockroach/blob/82b417793aade4fc1396524b184ed25ceaa32326/pkg/sql/ttl/ttljob/ttljob_processor.go#L463)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: `Execute` may be more clear than `Start` for the function name.
</blockquote></details>

That's a good idea. Done.
___
*[`pkg/util/retry/batch.go` line 19 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141953#-OK8MJFU-oMbHP85MPUT:-OKBczMa-K4xNXvK-8BO:b-20oiy6) ([raw file](https://github.com/cockroachdb/cockroach/blob/82b417793aade4fc1396524b184ed25ceaa32326/pkg/util/retry/batch.go#L19)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this is a nice utility! i think the audience in #db-engineering would be interested to know about it. up to you if you'd like to solicit broader feedback on it before merging, or if you'd like to just announce this later on.
</blockquote></details>

Cool. I'll announce it after merging. I also want to convert the other two cases where I added the retry (backfill and merge) to use retry.Batch before the announcement, so that we can ensure it's generic enough.
___
*[`pkg/util/retry/batch.go` line 26 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141953#-OK8KZXF2-Wxb_T2FaiQ:-OKBgDvODWGOdYs8wiXN:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/82b417793aade4fc1396524b184ed25ceaa32326/pkg/util/retry/batch.go#L26)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: could these fields just be `int`s?
</blockquote></details>

Done
___
*[`pkg/util/retry/batch.go` line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141953#-OK8K3ijB4K1HWcciJoc:-OKBhS_jEx9PdtClDaa0:bvcroz8) ([raw file](https://github.com/cockroachdb/cockroach/blob/82b417793aade4fc1396524b184ed25ceaa32326/pkg/util/retry/batch.go#L73)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i believe this line means that IsRetriable is not optional, but we might want to clarify the comment above that says users "can optionally define retry behavior using `IsRetriableError` and `OnRetry`"
</blockquote></details>

Good catch. I corrected the comment and added an assert to make sure `IsRetriableError` is provided.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141953)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/141953#-OK8N4PKBoPRkmN97Yx0:-OKXzXv8C7vzZ2bQ0cX_:b-d466ln) (no related file):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nice! this lgtm, but my main question is about adding a TTL-specific test.
</blockquote></details>

the new test and mock builders are very nice. lgtm!
___
*[`pkg/sql/ttl/ttljob/ttljob_processor_internal_test.go` line 232 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141953#-OKXzFr8DXpAZhs6LOmT:-OKXzFr97XCxUnMCiZo_:b-loo1pk) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bf9c28d3698d7dcc03349d35d1ed27d65ef6bc8/pkg/sql/ttl/ttljob/ttljob_processor_internal_test.go#L232)):*
> ```Go
> 			}
> 			// Check the metric to ensure we actually did the expected number of retries.
> 			require.Equal(t, metrics.NumDeleteBatchRetries.Value(), tc.expectedRetryCount)
> ```

super nit: the expected value should be the 2nd argument, and the actual value should be the 3rd


<!-- Sent from Reviewable.io -->

rafiss commented:
I added backport labels, since we are actively seeing this issue occur in a few Cloud clusters.

spilchen committed:
sql/ttl: add retry logic for TTL deletions

Previously, when a TTL deletion encountered contention, the KV layer
would automatically retry up to 100 times. If all 100 retries failed,
the error would propagate up the stack, causing the TTL deletion job to
restart.

A problem with this approach is that it will retry the TTL deletion
using the same batch size, making it likely that it will hit a retry
again.

This change introduces retry logic within the TTL deletion job itself.
If the KV auto-retry limit is reached, the delete batch size is halved
and retried. This process continues until the batch size reaches a
single row, at which point it will fail if the operation still cannot
succeed.

By dynamically adjusting the batch size instead of failing outright,
this change improves the robustness of TTL-based row deletions, reducing
the likelihood of job failures that would otherwise require a full
restart.

To support this, a new `retry.Batch` struct was added to the `retry`
package. In a follow-up PR, this struct will be used to replace similar
retry logic in other areas, such as the backfill and merge phases.

A new metric called jobs.row_level_ttl.num_delete_batch_retries has been
added to track the number of times the batch size has been reduced and
retried.

Epic: None
Release note (bug fix): The TTL deletion job now includes a retry
mechanism that progressively reduces the batch size when encountering
contention. This improves the chances of successful deletion without
requiring manual adjustments to TTL knobs.
Release note (ops change): Add jobs.row_level_ttl.num_delete_batch_retries
to track the number of times the TTL job had to reduce the batch size
and try again.
Closes #125103

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141953)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/ttl/ttljob/ttljob_processor_internal_test.go` line 232 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141953#-OKXzFr8DXpAZhs6LOmT:-OKYvhY5EaD3GBSVuf1H:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bf9c28d3698d7dcc03349d35d1ed27d65ef6bc8/pkg/sql/ttl/ttljob/ttljob_processor_internal_test.go#L232)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: the expected value should be the 2nd argument, and the actual value should be the 3rd
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR!

bors r+
